wp-includes/rest-api/auth/class-wp-rest-token.php

```diff
@@ -0,0 +1,703 @@
+		$days = apply_filters( 'rest_authentication_token_expire_days', 7 );
+		$time = time();
+
+		// JWT Reserved claims.
+		$reserved = array(
+			'iss' => get_bloginfo( 'url' ), // Token issuer.
+			'iat' => $time, // Token issued at.
+			'nbf' => $time, // Token accepted not before.
+			'exp' => $time + ( DAY_IN_SECONDS * absint( $days ) ), // Token expiry.
+		);
+
+		// JWT Private claims.
+		$private = array(
+			'data' => array(
+				'user' => array(
+					'id'         => $user->ID,
+					'type'       => isset( $user->type ) ? $user->type : 'wp_user',
+					'user_login' => $user->data->user_login,
+					'user_email' => $user->data->user_email,
+				),
+			),
+		);
+
+		/**
+		 * JWT Payload.
+		 *
+		 * The `data` private claim will always be added, but additional claims can be added via the
+		 * `rest_authentication_token_private_claims` filter. The data array will be included in the
+		 * REST response, do not include sensitive user data in that array.
+		 *
+		 * @param array           $payload The payload used to generate the token.
+		 * @param WP_User|Object  $user The authenticated user object.
+		 * @param WP_REST_Request $request The authentication request.
+		 */
+		$payload = apply_filters(
+			'rest_authentication_token_private_claims',
+			array_merge( $reserved, $private ),
+			$user,
+			$request
+		);
+
+		// Generate JWT token.
+		$token = $this->jwt( 'encode', $payload, $this->secret_key );
+
+		/**
+		 * Return response containing the JWT token and $user data.
+		 *
+		 * @param array           $response The REST response.
+		 * @param WP_User|Object  $user The authenticated user object.
+		 * @param WP_REST_Request $request The authentication request.
+		 */
+		return apply_filters(
+			'rest_authentication_token_response',
+			array(
+				'access_token' => $token,
+				'data'         => $payload['data'],
+			),
+			$user,
+			$request
+		);
+	}
+
+	/**
+	 * Determine if a valid Bearer token has been provided.
+	 *
+	 * @return object|WP_Error Return the JSON Web Token object, or WP_Error on failure.
+	 */
+	public function validate_token() {
+
+		// Get HTTP Authorization Header.
+		$header = $this->get_auth_header();
+		if ( is_wp_error( $header ) ) {
+			return $header;
+		}
+
+		// Get the Bearer token from the header.
+		$token = $this->get_token( $header );
+		if ( is_wp_error( $token ) ) {
+			return $token;
+		}
+
+		// Decode and validate the access token.
+		try {
+
+			// Decode the token.
+			$jwt = $this->jwt( 'decode', $token, $this->secret_key, array( 'HS256' ) );
+
+			// Determine if the token issuer is valid.
+			$issuer_valid = $this->validate_issuer( $jwt->iss );
+			if ( is_wp_error( $issuer_valid ) ) {
+				return $issuer_valid;
+			}
+
+			// Determine if the token user is valid.
+			$user_valid = $this->validate_user( $jwt );
+			if ( is_wp_error( $user_valid ) ) {
+				return $user_valid;
+			}
+
+			// Determine if the token has expired.
+			$expiration_valid = $this->validate_expiration( $jwt );
```

Comment:
This should never be reached if the token has expired because [`JWT::decode()` throws an exception](https://github.com/firebase/php-jwt/blob/v5.0.0/src/JWT.php#L115-L135) in this case.

We should probably catch the specific errors `\Firebase\JWT\BeforeValidException` and `\Firebase\JWT\ExpiredException` before falling back to the generic `\Exception`.